components/suggestion/at_mention_provider/at_mention_provider.jsx

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import XRegExp from 'xregexp';
+
+import {Constants} from 'utils/constants.jsx';
+
+import Provider from '../provider.jsx';
+
+import AtMentionSuggestion from './at_mention_suggestion.jsx';
+
+// The AtMentionProvider provides matches for at mentions, including @here, @channel, @all,
+// users in the channel and users not in the channel. It mixes together results from the local
+// store with results fetch from the server.
+export default class AtMentionProvider extends Provider {
+    constructor(props) {
+        super();
+
+        this.setProps(props);
+
+        this.data = null;
+    }
+
+    // setProps gives the provider additional context for matching pretexts. Ideally this would
+    // just be something akin to a connected component with access to the store itself.
```

Comment:
This is something I've wanted to look at as well. We could either:
1. Change the Provider classes to components that just render nothing so that we can render something like `<SuggestionBox><AtMentionProvider/></SuggestionBox>`
2. Hook into the context when instantiating the `Textbox` to pass the store directly to the Provider

The first method is a bit hacky since the provider is not really a component, but it is cleaner than the second one since that requires us to breach the encapsulation provided by `react-redux`

Reply:
The first also naturally moves us towards triggering a "render" (i.e. prefix match) on any change, while the latter has the difficulty of still needing to notify the renderer that the provider has changed, perhaps outside of an explicit pretext match. Complicated either way, I suppose!